File: Scala-Types/src/main/scala/ch/mse/app/types/path/Outer.scala
package ch.mse.app.types.path

class Outer {
  class Inner
  def put(inner: this.Inner) {}
}

object Outer {
  val o1 = new Outer
  val o2 = new Outer

  def foo(x: o1.Inner) = 3
//  def foo(x: o2.Inner) = 4 // problem due to erasure
  
  
  def main(args: Array[String]) {
    
    val i1 = new o1.Inner
    val i2 = new o2.Inner
    
    foo(i1);
    // foo(i2); // type mismatch
    
  }
}

File: Scala-Types/src/main/java/ch/mse/app/types/Outer.java
package ch.mse.app.types;

public class Outer {

	class Inner {
	}

	public void put(Inner inner) {

	}

	public static void main(String[] args) {
		Outer outer1 = new Outer();
		Inner inner1 = outer1.new Inner();

		Outer outer2 = new Outer();
		Inner inner2 = outer2.new Inner();

		outer1.put(inner2);
		outer2.put(inner1);
	}
}
